{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04632f9e-b8db-407f-a2a8-fb7835a6f4ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "from abc import ABC, abstractmethod\n",
    "from Bio.Data import CodonTable\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b8c728-9295-48b5-9010-b66343504a03",
   "metadata": {},
   "source": [
    "This notebook is for putting together the end-to-end tests of evSeq. We will reconstruct inputs for known outputs, then test them make sure the evSeq software outputs what we expect.\n",
    "\n",
    "Steps:\n",
    "1. Build a random amino acid sequence. This is our reference AA.\n",
    "2. Decide if we are operating in detailed_refseq mode or not. If not, then the random amino acid sequence gives what we expect for the full plate.\n",
    "3. Build variants of the input amino acid sequences. Decide if we have a mixed well or not. A mixed well means we need a second amino acid sequence.\n",
    "4. Build both a \"NNN\" and no variation version of the reference sequence corresponding to positions where we expect mutations.\n",
    "5. Choose a random codon to encode the amino acid changes in the sequence.\n",
    "6. For building reads, choose:\n",
    "    1. Ratio of different sequences. This decides whether or not they are recognized by `variable_thresh`\n",
    "    2. Number of total sequences *at each position*. This decides whether a well is considered dead or not by `variable_count`\n",
    "7. Build a set of perfect reads up to the total number of target sequences. \n",
    "8. Add on bad sequences that should get filtered out by the `bp_q_cutoff`, `length_cutoff`, and `average_q_cutoff` parameters. Also add indels.\n",
    "9. Pad on additional sequence information that creates frameshifts. Add to an input file for processing."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f493de-df68-433f-842c-6ace278fbfa1",
   "metadata": {},
   "source": [
    "Some methods are shared. These are held in an abstract base class:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d572223-75de-4af9-9a6f-10a72b0c8f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FakeData(ABC):\n",
    "    \"\"\"\n",
    "    Abstract class holding shared methods between data generators for testing.\n",
    "    \"\"\"\n",
    "    @abstractmethod\n",
    "    def build_output_counts(self):\n",
    "        \"\"\"\n",
    "        Builds output files for the different `OutputCounts`\n",
    "        \"\"\"\n",
    "        pass\n",
    "        \n",
    "    @abstractmethod\n",
    "    def build_filtered_fastqs(self):\n",
    "        \"\"\"\n",
    "        Builds the ParsedFilteredFastqs files for the different wells.\n",
    "        \"\"\"\n",
    "        # Use the `perfect_reads` + `corrupted_bases` reads to generate what we expect\n",
    "        # for the evSeq `ParsedFilteredFastqs` output\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e930105f-b8e6-444f-a108-c604aabaead2",
   "metadata": {},
   "source": [
    "The BioPython codon table implementation is beyond bad, so we rewrite it here, fixing all of the BioPython flaws:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6e03e4ac-d2ac-413d-a41a-c09ae1d6d313",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CustomCodonTable():\n",
    "    \"\"\"\n",
    "    Because the BioPython implementation of a codon table is all over the place,\n",
    "    fixing all the flaws in this class. Not inheriting because I just don't trust it.\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        \n",
    "        # Get the standard codon table from BioPython. Their docs do not say what this is,\n",
    "        # but it looks right from a spot check. As it is, we wrote our own for evSeq, so it\n",
    "        # is good to have a separate codon table used for validation.\n",
    "        bio_codon_table = CodonTable.standard_dna_table\n",
    "        \n",
    "        # Build the forward dictionary. Add the stop codon translations (why\n",
    "        # aren't these included in the first place???).\n",
    "        self.codon_to_aa = bio_codon_table.forward_table.copy()\n",
    "        for stop_codon in bio_codon_table.stop_codons:\n",
    "            self.codon_to_aa[stop_codon] = \"*\"\n",
    "            \n",
    "        # Build the reverse dictionary. BioPython reverse dictionary includes just one \n",
    "        # codon per aa. Fix this.\n",
    "        self.aa_to_codon = {}\n",
    "        for codon, aa in self.codon_to_aa.items():\n",
    "            if aa in self.aa_to_codon:\n",
    "                self.aa_to_codon[aa].append(codon)\n",
    "            else:\n",
    "                self.aa_to_codon[aa] = [codon]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b7ea83a-31ae-4f6a-8755-4838f738a79d",
   "metadata": {},
   "source": [
    "We need global variables that define the bounds on our testing. Everything is randomly produced, so we need to set ranges on our sampling:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0a245856-42f2-497a-8050-9e8265830bab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import evSeq globals needed for testing\n",
    "from evSeq.util.globals import (BARCODE_LENGTH, \n",
    "                                ADAPTER_F,\n",
    "                                ADAPTER_LENGTH_F,\n",
    "                                ADAPTER_R,\n",
    "                                ADAPTER_LENGTH_R)\n",
    "\n",
    "# Reference sequence bounds\n",
    "MIN_REFSEQ_LEN = 150\n",
    "MAX_REFSEQ_LEN = 700\n",
    "\n",
    "# Bounds on readlength\n",
    "MIN_READLENGTH = 150\n",
    "MAX_READLENGTH = 300\n",
    "\n",
    "# Bounds on quality cutoffs\n",
    "MIN_BP_QUAL_CUTOFF = 15\n",
    "MAX_BP_QUAL_CUTOFF = 35\n",
    "\n",
    "MIN_GLOBAL_QUAL_CUTOFF = 15\n",
    "MAX_GLOBAL_QUAL_CUTOFF = 35\n",
    "\n",
    "# Bounds on sequence length cutoffs\n",
    "MIN_SEQLEN_CUTOFF = 0.0\n",
    "MAX_SEQLEN_CUTOFF = 1.0\n",
    "\n",
    "# Bounds on number of indels added\n",
    "MIN_INDELS_ADDED = 1\n",
    "MAX_INDELS_ADDED = 3\n",
    "\n",
    "# Bounds on primer lengths\n",
    "PRIMER_MIN_LEN = 10\n",
    "PRIMER_MAX_LEN = 40\n",
    "\n",
    "# Bounds on the frameshift for a reference seq\n",
    "FRAMESHIFT_MIN = 0\n",
    "FRAMESHIFT_MAX = 3\n",
    "\n",
    "# Codon table and allowed amino acid characters\n",
    "CODON_TABLE = CustomCodonTable()\n",
    "ALLOWED_AAS = tuple(sorted(list(CODON_TABLE.aa_to_codon.keys())))\n",
    "INT_TO_AA = dict(enumerate(ALLOWED_AAS))\n",
    "N_AAS = len(ALLOWED_AAS)\n",
    "\n",
    "# Random number generators\n",
    "RANDOM_SEED = sum(ord(char) for char in \"PDawg\")\n",
    "NP_RNG = np.random.default_rng(RANDOM_SEED)\n",
    "RANDOM_RNG = random.Random(RANDOM_SEED)\n",
    "\n",
    "# Barcode file\n",
    "INDEX_DF = pd.read_csv(\"../evSeq/util/index_map.csv\")\n",
    "N_INDICES = len(INDEX_DF)\n",
    "N_PLATES = len(INDEX_DF.IndexPlate.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01776e8e-6cd9-4c7c-a11e-382fb8f92d5a",
   "metadata": {},
   "source": [
    "We will store basic variables used for building test data in a configuration class, defined below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8192e0be-4da8-409e-96e6-ea53c56ed522",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FakeRefseq():\n",
    "    def __init__(self):\n",
    "        \n",
    "        # Randomly create an amino acid reference sequence. Record the length.\n",
    "        self.refseq_len = NP_RNG.integers(MIN_REFSEQ_LEN, MAX_REFSEQ_LEN)\n",
    "        aa_ints = NP_RNG.choice(N_AAS, size = self.refseq_len)\n",
    "        self.aa_refseq = [INT_TO_AA[aa_int] for aa_int in aa_ints]\n",
    "\n",
    "        # Assign a codon to each amino acid.\n",
    "        self.codon_refseq = [RANDOM_RNG.choice(CODON_TABLE.aa_to_codon[aa]) \n",
    "                             for i, aa in enumerate(self.aa_refseq)] \n",
    "        \n",
    "        # Decide on the frameshift of the reference sequence\n",
    "        self.frameshift_front = NP_RNG.integers(FRAMESHIFT_MIN, FRAMESHIFT_MAX)\n",
    "        self.frameshift_back = NP_RNG.integers(FRAMESHIFT_MIN, FRAMESHIFT_MAX)        \n",
    "        \n",
    "        # Assign a length to the codon reference sequence\n",
    "        self.codon_refseq_len = (self.refseq_len * 3 + \n",
    "                                 self.frameshift_front +\n",
    "                                 self.frameshift_back)\n",
    "        \n",
    "        # Assign the primer seed length\n",
    "        self.primer_seed_len = NP_RNG.integers(PRIMER_MIN_LEN, PRIMER_MAX_LEN)\n",
    "    \n",
    "    def define_refseq_windows(self, readlength):\n",
    "        \"\"\"\n",
    "        Defines the primers for the refseq as well as the variable window\n",
    "        \"\"\"\n",
    "        # Choose legal positions for making mutations. This is the last amino acid\n",
    "        # captured in full by the readlength, considering the primer binding region,\n",
    "        # the adapter machinery, and the barcode.\n",
    "        effective_readlength = readlength - BARCODE_LENGTH\n",
    "        max_readable_aa_ind_f = (effective_readlength - self.frameshift_front - ADAPTER_LENGTH_F) // 3\n",
    "        min_readable_aa_ind_r = (effective_readlength - self.frameshift_back - ADAPTER_LENGTH_R) // 3\n",
    "        \n",
    "        aa_refseq_len = len(self.aa_refseq)\n",
    "        mutable_aa_inds = (list(range(0, max_readable_aa_ind_f)) + \n",
    "                           list(range(aa_refseq_len - min_readable_aa_ind_r, aa_refseq_len)))\n",
    "        \n",
    "        # Get only unique mutable inds\n",
    "        self.mutable_aa_inds = np.array(list(set(mutable_aa_inds)), dtype = int)\n",
    "        self.mutable_aa_inds.sort()\n",
    "        \n",
    "# Class that holds parameters that will be needed by evSeq\n",
    "class Config():\n",
    "    def __init__(self, detailed = True):\n",
    "        \"\"\"\n",
    "        Builds a set of conditions that might be passed into an evSeq run.\n",
    "        \"\"\"\n",
    "        # Build as many reference sequences as needed\n",
    "        n_refs_needed = N_INDICES if detailed else N_PLATES\n",
    "        self.refseqs = [FakeRefseq() for _ in range(n_refs_needed)] \n",
    "        \n",
    "        # Decide on the readlength that we will be using. \n",
    "        min_refseq_len = min(refseq.codon_refseq_len for refseq in self.refseqs)                             \n",
    "        min_readlength = min(MIN_READLENGTH, min_refseq_len)\n",
    "        self.readlength = NP_RNG.integers(min_readlength, MAX_READLENGTH)\n",
    "        \n",
    "        # Define the allowed mutagenesis windows for the amino acids. This is the \n",
    "        # last amino acid captured in full by the readlength\n",
    "        for refseq in self.refseqs:\n",
    "            refseq.define_refseq_windows(self.readlength)\n",
    "        \n",
    "        # Decide on the input variables that will define the run, including \n",
    "        # average_q_cutoff, bp_q_cutoff and length_cutoff\n",
    "        self.average_q_cutoff = NP_RNG.integers(MIN_GLOBAL_QUAL_CUTOFF, MAX_GLOBAL_QUAL_CUTOFF)\n",
    "        self.bp_q_cutoff = NP_RNG.integers(MIN_BP_QUAL_CUTOFF, MAX_BP_QUAL_CUTOFF)\n",
    "        self.length_cutoff = NP_RNG.uniform(MIN_SEQLEN_CUTOFF, MAX_SEQLEN_CUTOFF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e77d83d1-54d6-4d6e-841e-ae1c432be02e",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_config = Config()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a41dcb37-fce2-4f65-b0c9-7908e16c7a1c",
   "metadata": {},
   "source": [
    "The basic data generator is a FakeWell instance. This holds methods for building fake data from the Config object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a981cd0-c862-4412-9bc3-dbabf7346185",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Class that holds information for a test well\n",
    "class FakeWell(FakeData):\n",
    "    def __init__(self, reference_sequence):\n",
    "    \n",
    "        # Assign a reference sequence object as an instance variable\n",
    "        self.refseq = reference_sequence\n",
    "        \n",
    "        # Create a barcode variable as a placeholder. This will be filled when the well\n",
    "        # is passed with others to a FakeRun instance.\n",
    "        self.f_barcode = None\n",
    "        self.r_barcode = None\n",
    "        \n",
    "    def choose_variable_positions(self):\n",
    "        pass\n",
    "        # Decide on the positions that will be varied within the sequence within the bounds\n",
    "        # chosen during the initialization of the class.\n",
    "        \n",
    "        # Create a \"NNN\" version of the reference sequence for use in processing.\n",
    "        # Note that we don't need to capture all positions -- we want evSeq to\n",
    "        # handle finding some of them sometimes\n",
    "        \n",
    "    def choose_variants(self):\n",
    "        pass\n",
    "        # Decide how many variants we want in the well\n",
    "        \n",
    "        # Choose variable amino acids.\n",
    "        \n",
    "        # Choose a random codon used for encoding amino acids\n",
    "        \n",
    "    def build_perfect_reads(self):\n",
    "        \"\"\"\n",
    "        Assigns a `perfect_reads` variable to the instance.\n",
    "        \"\"\"\n",
    "        pass\n",
    "        # Decide on the ratio between the different variants\n",
    "        \n",
    "        # Get the total sequences we want at each **position**\n",
    "        \n",
    "        # Build a set of perfect reads up to the target numbers\n",
    "        \n",
    "    def build_corrupted_reads(self):\n",
    "        \"\"\"\n",
    "        Assigns `corrupted_reads` and `corrupted_bases` variables to the instance.\n",
    "        \"\"\"\n",
    "        pass\n",
    "        # Choose how many sequences to add of each flavor.\n",
    "        \n",
    "        # Add sequences filtered out by bp_q_cutoff. This is `corrupted_bases`\n",
    "        \n",
    "        # Add indels. These will go into `corrupted_bases`. We need to record the positions \n",
    "        # where indels are found.\n",
    "        \n",
    "        # Add sequences filtered out by the average_q_cutoff. This is `corrupted_reads`.\n",
    "        \n",
    "        # Add sequences filtered out by the length_cutoff. This is added to `corrupted_reads`.\n",
    "                \n",
    "                        \n",
    "    def pad_for_frameshift(self):\n",
    "        pass\n",
    "        # Adds bases to the beginning and end of the sequences to cause frameshifts.\n",
    "        # These bases will be analyzed by the software, but should not affect AA indices.\n",
    "        \n",
    "    def build_output_counts(self):\n",
    "        \"\"\"\n",
    "        Builds output files for the different `OutputCounts`\n",
    "        \"\"\"\n",
    "        pass\n",
    "        \n",
    "    def build_filtered_fastqs(self):\n",
    "        \"\"\"\n",
    "        Builds the ParsedFilteredFastqs files for the different wells.\n",
    "        \"\"\"\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a23c3e0b-3264-43f0-a0dd-09e078c0add8",
   "metadata": {},
   "source": [
    "All FakeWell instances are held within a FakeRun instance that assigns barcodes to the fake wells, handles fake input generation, and builds expected outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1c7e3758-0bdf-4968-9a86-8506ed7bf485",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Class that holds information for a test run\n",
    "def FakeRun(FakeData):\n",
    "    def __init__(self, fakewells, detailed = True):\n",
    "        \"\"\"\n",
    "        fakewells: A list of fully prepared FakeWell objects.\n",
    "        detailed: Whether or not we are using a detailed refseq file. \n",
    "        \"\"\"\n",
    "        # Assign fakewells as an instance variable\n",
    "        pass\n",
    "    \n",
    "    def build_input(self):\n",
    "        pass\n",
    "        # Assign a barcode combo to each of the fakewells\n",
    "        \n",
    "        # Build fastq files from the fakewells\n",
    "        \n",
    "        # Return the fastq files ready for processing\n",
    "        \n",
    "    def build_output_counts(self):\n",
    "        \"\"\"\n",
    "        Builds output files for the different `OutputCounts`\n",
    "        \"\"\"\n",
    "        pass\n",
    "        \n",
    "    def build_filtered_fastqs(self):\n",
    "        \"\"\"\n",
    "        Builds the ParsedFilteredFastqs files for the different wells.\n",
    "        \"\"\"\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e17e1fd2-8327-4d57-9e1d-f9e4d389d6d5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
